finetrainers/trainer.py

```diff
@@ -0,0 +1,542 @@
+import inspect
+import json
+import logging
+import math
+import os
+import random
+import shutil
+from datetime import timedelta
+from typing import Any, Dict
+from pathlib import Path
+
+import diffusers
+import torch
+import torch.backends
+import transformers
+from accelerate import Accelerator, DistributedType
+from accelerate.logging import get_logger
+from accelerate.utils import DistributedDataParallelKwargs, InitProcessGroupKwargs, ProjectConfiguration, set_seed
+from diffusers.optimization import get_scheduler
+from diffusers.training_utils import cast_training_params, compute_density_for_timestep_sampling, compute_loss_weighting_for_sd3
+from huggingface_hub import create_repo, upload_folder
+from peft import LoraConfig, get_peft_model_state_dict, set_peft_model_state_dict
+from tqdm import tqdm
+
+from .args import Args
+from .constants import FINETRAINERS_LOG_LEVEL
+from .dataset import BucketSampler, VideoDatasetWithResizing
+from .models import get_config_from_model_name
+from .state import State
+from .utils.optimizer_utils import get_optimizer, gradient_norm
+from .utils.memory_utils import bytes_to_gigabytes, make_contiguous
+from .utils.torch_utils import unwrap_model
+
+
+logger = get_logger("finetrainers")
+logger.setLevel(FINETRAINERS_LOG_LEVEL)
+
+class Trainer:
```

Comment:
This could provide util methods just like our Mixin classes do and we could make methods like `prepare_models()` abstract methods and raise a `NotImplementedError` as needed. 

This way we delegate the abstractions better, IMO, and get rid of nasty if/else blocks.

Reply:
Yes, that sounds good. I would like to refactor this and make it easier to use, but for this PR, let's roll with sole focus on LTXV. Once I get to Hunyuan and Mochi, it will make it easier to find what needs abstraction and how best to do it. Also yes, providing abstractions for others to implement their own custom methods that are not part of our codebase sounds good

Reply:
Sure. I can take either of Mochi and Hunyuan to free your plate a bit.

Reply:
Awesome! Would you like to do Mochi since I have already started on Hunyuan locally?

Reply:
Leave it with me. Over n' out.